### PR TITLE
WIP: GHC 8.4 (singletons-2.4)

### LINKED ISSUE #254: GHC 8.4 checklist
GHC 8.4 adds some extra goodies that would be nice to incorporate into `singletons`. This issue serves as a place to record all of these features as a reminder to fold them in before cutting an 8.4-compatible release.

* [ ] There are now [built-in `Div` and `Mod` type families](http://git.haskell.org/ghc.git/commit/fa8035e3ee83aff5a20fc5e7e2697bac1686d6a6) in `GHC.TypeNats` that use efficient machine operations under the hood. We can use these to vastly speed up the current `PShow` instance for `Nat`.

    While we're in town, we ought to work `singletons`' magic on `Log2` as well.
* [ ] Due to [recent work on data families](https://ghc.haskell.org/trac/ghc/ticket/12369), it is now possible to write `Sing` without extraneous type variables, like so:

  ```haskell
  data family Sing :: k -> Type
  data instance Sing :: [a] -> Type where ...
  ```

  As opposed to before, where we had to conjure up unused type variables to work around GHC restrictions:

  ```haskell
  data family Sing (a :: k)
  data instance Sing (z :: [a]) where ...
  ```

  In order to apply this change to its fullest, we will first need to update `th-desugar` to allow GADTs with explicit return kinds. See https://github.com/goldfirere/th-desugar/issues/68.

### PR DESCRIPTION
**Pull request by RyanGlScott**
This provides the necessary changes to the `singletons` library to support GHC 8.4. They're fairly minimal this go around. They include:

* Adding `MonadIO` instances for some newtypes, since `MonadIO` is now a superclass of `Quasi`.
* Implementing `qAddCorePlugin` for `Quasi` instances.
* Turning instances of the form `data instance Sing (z :: Foo)` to `data instance Sing :: Foo -> Type` where possible (see #254).
* Using the built-in type families `Div`, `Mod`, and `Log2` from `GHC.TypeNats` within `Data.Singletons.TypeLits` (see #254).

Fixes #254.

There's one remain snag: the Travis build won't currently work, since not all dependencies from Hackage currently build with GHC 8.4.1. We could conceivable work around this with `head.hackage`, but it would require a non-trivial amount of effort to make `singletons`' test suite support `cabal new-build` (see #225).

### CODE REVIEW

File: src/Data/Singletons/TypeLits.hs [hunk 1]
```diff
@@ -98,32 +100,63 @@ no_term_level_syms = error "The kind `Symbol` may not be used at the term level.
...
-
-  divMod' :: Nat -> Nat -> Nat -> Nat -> (Nat, Nat)
-  divMod' 0 _ q u = (q, u)
-  divMod' n y q 0 = divMod' (n-1) y (q+1) y
-  divMod' n y q u = divMod' (n-1) y q     (u-1)
-
-  div :: Nat -> Nat -> Nat
-  div _ 0 = error "Division by zero."
-  div x y = fst (divMod x y)
-  infixl 7 `div`
```
Comment (RyanGlScott) [inline src/Data/Singletons/TypeLits.hs]: One question you might ask is: do the built-in `Div` and `Mod` type families from `GHC.TypeNats` have the same fixity (`infixl 7`)? Sadly, the answer is currently "no", although I'm seeking to correct this discrepancy in [GHC Trac #14640](https://ghc.haskell.org/trac/ghc/ticket/14640).

File: CHANGES.md [hunk 2]
```diff
@@ -1,8 +1,8 @@
 Changelog for singletons project
 ================================
 
-next
-----
+2.4
```
Comment (goldfirere) [inline CHANGES.md]: This should mention that we're switching support to 8.4.

File: singletons.cabal [hunk 3]
```diff
@@ -42,7 +42,7 @@ source-repository this
 
 library
   hs-source-dirs:     src
-  build-depends:      base >= 4.10 && < 5,
+  build-depends:      base >= 4.11 && < 5,
```
Comment (goldfirere) [inline singletons.cabal]: Should we have 4.12 as an upper bound?

Comment (RyanGlScott) [reply singletons.cabal]: Good idea, fixed.

**Comment by goldfirere** [comment]
To be clear, please merge when ready.

**Pull request merged**